python-checks/src/main/java/org/sonar/python/checks/StrftimeConfusingHourSystemCheck.java

```diff
@@ -0,0 +1,93 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.checks.utils.Expressions;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6883")
+public class StrftimeConfusingHourSystemCheck extends PythonSubscriptionCheck {
+
+  public static final String MESSAGE = "Use %I (12-hour clock) or %H (24-hour clock) without %p (AM/PM).";
+  public static final String MESSAGE_12_HOURS = "Use %I (12-hour clock) with %p (AM/PM).";
+  private static final String MESSAGE_SECONDARY_LOCATION = "Wrong format created here.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, StrftimeConfusingHourSystemCheck::checkCallExpr);
+  }
+
+  private static void checkExpression(SubscriptionContext context, Expression expression) {
+    checkExpression(context, expression, expression);
+  }
+
+  private static void checkExpression(SubscriptionContext context, Expression expression, Tree primaryLocation) {
+    if (expression.is(Tree.Kind.NAME)) {
+      Expressions.singleAssignedNonNameValue((Name) expression).ifPresent(a -> checkExpression(context, a, primaryLocation));
+    }
+    if (expression.is(Tree.Kind.STRING_LITERAL)) {
+      StringLiteral stringLiteral = (StringLiteral) expression;
+      if (stringLiteral.stringElements().stream().anyMatch(stringElement -> "f".equalsIgnoreCase(stringElement.prefix()))) {
+        return;
+      }
+      String value = stringLiteral.trimmedQuotesValue();
+      String effectiveMessage = null;
+      if (value.contains("%H") && value.contains("%p")) {
+        effectiveMessage = MESSAGE;
+      } else if (value.contains("%I") && !value.contains("%p")) {
+        effectiveMessage = MESSAGE_12_HOURS;
+      }
+      if (effectiveMessage != null) {
+        var issue = context.addIssue(primaryLocation, effectiveMessage);
+        if (primaryLocation != expression) {
+          issue.secondary(stringLiteral, MESSAGE_SECONDARY_LOCATION);
+        }
+      }
+    }
+  }
```

Comment:
This could be implemented in the following way:
```java
  private static void checkExpression(SubscriptionContext context, Expression expression, Tree primaryLocation) {
    Expressions.ifNameGetSingleAssignedNonNameValue(expression)
      .filter(StringLiteral.class::isInstance)
      .map(StringLiteral.class::cast)
      .filter(stringLiteral -> stringLiteral.stringElements().stream().noneMatch(stringElement -> "f".equalsIgnoreCase(stringElement.prefix())))
      .ifPresent(stringLiteral -> checkDateFormatStringLiteral(context, primaryLocation, stringLiteral));
  }

  private static void checkDateFormatStringLiteral(SubscriptionContext context, Tree primaryLocation, StringLiteral stringLiteral) {
    String value = stringLiteral.trimmedQuotesValue();
    String effectiveMessage = null;
    if (value.contains("%H") && value.contains("%p")) {
      effectiveMessage = MESSAGE;
    } else if (value.contains("%I") && !value.contains("%p")) {
      effectiveMessage = MESSAGE_12_HOURS;
    }
    if (effectiveMessage != null) {
      var issue = context.addIssue(primaryLocation, effectiveMessage);
      if (primaryLocation != stringLiteral) {
        issue.secondary(stringLiteral, MESSAGE_SECONDARY_LOCATION);
      }
    }
  }
```